core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -242,85 +246,116 @@ final class KafkaMetadataLog private (
   }
 
   override def readSnapshot(snapshotId: OffsetAndEpoch): Optional[RawSnapshotReader] = {
-    try {
-      if (snapshotIds.contains(snapshotId)) {
-        Optional.of(FileRawSnapshotReader.open(log.dir.toPath, snapshotId))
-      } else {
-        Optional.empty()
+    snapshots synchronized {
+      val reader = snapshots.get(snapshotId) match {
+        case None =>
+          // Snapshot doesn't exists
+          None
+        case Some(None) =>
+          // Snapshot exists but has never been read before
+          try {
+            val snapshotReader = Some(FileRawSnapshotReader.open(log.dir.toPath, snapshotId))
+            snapshots.put(snapshotId, snapshotReader)
+            snapshotReader
```

Comment:
Just to clarify my understanding -- this is the main change in this PR right? The FileRawSnapshotReader-s are now owned (opened and closed) by KafkaMetadataLog whereas previously they were opened here, but never closed. 

Now in KafkaMetadataLog#close and KafkaMetadataLog#deleteSnapshotFiles, we are closing the FileRawSnapshotReader instances that we opened.

Reply:
Correct.